lib/jormungandr/src/Cardano/Wallet/Jormungandr/Transaction.hs

```diff
@@ -63,39 +58,35 @@ import Data.Text.Class
 import Fmt
     ( Buildable (..) )
 
-import qualified Cardano.Crypto.Wallet as CC
-import qualified Cardano.Wallet.Jormungandr.Binary as Binary
 import qualified Data.ByteString.Lazy as BL
 
 -- | Construct a 'TransactionLayer' compatible with Shelley and 'Jörmungandr'
 newTransactionLayer
-    :: forall n k t.
+    :: forall k t.
         ( t ~ Jormungandr
-        , PaymentAddress n k
-        , MkTxWitness k
+        , TxWitnessTagFor k
         , WalletKey k
         )
     => Hash "Genesis"
     -> TransactionLayer t k
-newTransactionLayer (Hash block0H) = TransactionLayer
+newTransactionLayer block0H = TransactionLayer
     { mkStdTx = \keyFrom rnps outs -> do
-        -- NOTE
-        -- For signing, we need to embed a hash of the transaction data
-        -- without the witnesses (since we don't yet have them!). In this sense,
-        -- this is a transaction id as Byron nodes or the http-bridge
-        -- defines them.
+        credentials <- forM rnps $ \(_, TxOut addr _) -> first getRawKey <$>
+            maybeToRight (ErrKeyNotFoundForAddress addr) (keyFrom addr)
         let inps = fmap (second coin) rnps
-        wits <- forM rnps $ \(_, TxOut addr _) -> do
-            xprv <- maybeToRight (ErrKeyNotFoundForAddress addr) (keyFrom addr)
-            let payload = block0H <> getHash (signData inps outs)
-            pure $ mkTxWitness (fst xprv) (sign payload xprv)
-        let tx = Tx
-                { txId = fragmentId inps outs wits
+        let fragment = putFragment
+                block0H
+                (zip inps credentials)
+                outs
+                (MkFragmentSimpleTransaction (txWitnessTagFor @k))
+        return
+            ( Tx
+                { txId = fragmentId fragment
```

Comment:
We could (like in https://github.com/input-output-hk/cardano-wallet/pull/1021#discussion_r349290864) let `putFragment` return `Hash "Tx", SealedTx` also, and we'd
1. Be able to remove `fragmentId`
2. No longer have to call `BS.drop 2`

Reply:
On that topic, I think we're missing tests for `fragmentId`.

We should test that `fragmentId` here and `getFragment` results in the same id, and that it matches `jcli`.

Reply:
> We should test that fragmentId here and getFragment results in the same id, and that it matches jcli.

I agree. Currently, if all our unit tests pass, we have no guarantee that ids are computed correctly, which is kinda problematic.

Reply:
> We could (like in #1021 (comment)) let putFragment return Hash "Tx", SealedTx also, and we'd
> 
>    Be able to remove fragmentId
>    No longer have to call BS.drop 2

I strongly disagree here. `putFragment` should return a `PutM` and nothing else. Returning something else makes it harder to test and less intuitive. 

Being able to split out `fragmentId` into a separate function is actually a good thing (single responsibility principle for function, more composable). 

> No longer have to call BS.drop 2

Sort of inelegant, but very much acceptable in the binary module IMO.